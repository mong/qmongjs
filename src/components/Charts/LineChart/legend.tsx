import React, { Dispatch, useRef, SetStateAction, useEffect } from "react";
import { ScaleOrdinal } from "d3";
import style from "./legend.module.css";
import {
  useResizeObserver,
  useLegendItemPosition,
  useTextWidth,
} from "../../../helpers/hooks";

interface LegendProps {
  offsetLeft: number;
  offsetTop: number;
  legendWidth: number;
  legendHeight: number;
  setLegendHeight: Dispatch<SetStateAction<number>>;
  lineColorScale: ScaleOrdinal<string, string, never>;
  legendLabels: string[];
  setHoveredLegend: Dispatch<SetStateAction<string | null>>;
  selectedLegends: string[] | [];
  setSelectedLegends: Dispatch<SetStateAction<string[] | []>>;
}
export const Legend: React.FC<LegendProps> = ({
  offsetLeft,
  offsetTop,
  legendWidth,
  legendLabels,
  lineColorScale,
  legendHeight,
  setLegendHeight,
  setHoveredLegend,
  selectedLegends,
  setSelectedLegends,
}) => {
  const legendRef = useRef<SVGGElement | null>(null);

  const legendH = Math.ceil(
    useResizeObserver(legendRef)?.contentRect.height ?? 0
  );
  const legendLabelsString = legendLabels.toString();

  useEffect(() => {
    if (legendRef.current) {
      setLegendHeight(30 + legendH);
    }
  }, [legendHeight, legendH, legendLabelsString, setLegendHeight]);

  const LI = legendLabels.map((d, i) => {
    return (
      <LegendItem
        key={`${d} ${lineColorScale(d)}${i}`}
        label={d}
        legendLabelsString={legendLabelsString}
        color={lineColorScale(d)}
        legendWidth={legendWidth}
        setHoveredLegend={setHoveredLegend}
        selectedLegends={selectedLegends}
        setSelectedLegends={setSelectedLegends}
      />
    );
  });

  return (
    <g ref={legendRef} transform={`translate(${offsetLeft}, ${offsetTop})`}>
      {LI}
    </g>
  );
};

interface LegendItemProps {
  label: string;
  color: string;
  legendWidth: number;
  setHoveredLegend: Dispatch<SetStateAction<string | null>>;
  selectedLegends: string[];
  setSelectedLegends: Dispatch<SetStateAction<string[] | []>>;
  legendLabelsString: string;
}

const LegendItem: React.FC<LegendItemProps> = ({
  label,
  color,
  legendWidth,
  setHoveredLegend,
  selectedLegends,
  setSelectedLegends,
  legendLabelsString,
}) => {
  const legendRef = useRef<SVGGElement | null>(null);
  const textRef = useRef<SVGTextElement | null>(null);
  const textWidth = useTextWidth(textRef.current, label, color, legendWidth);

  const vMargin = 30;
  const hMargin = 10;

  const { x, y } = useLegendItemPosition(
    legendRef.current,
    legendWidth,
    vMargin,
    hMargin,
    legendLabelsString,
    textWidth
  );

  return (
    <g
      onMouseOver={() => setHoveredLegend(label)}
      onMouseOut={() =>
        setHoveredLegend((current: string | null) =>
          current === label ? null : current
        )
      }
      onClick={() => {
        setHoveredLegend((current) => (current === label ? null : current));
        setSelectedLegends((selected: string[]) => {
          if (selected.includes(label)) {
            return selected.filter((s) => s !== label);
          }
          return [...selected, label];
        });
      }}
      opacity={
        selectedLegends.length === 0
          ? 1
          : selectedLegends.includes(label)
          ? 1
          : 0.4
      }
      className={`${style.legendItem} ${label.replace(/\s/g, "")}`}
      ref={legendRef}
      transform={`translate(${x},${y})`}
    >
      <text
        ref={textRef}
        fontFamily="Arial, Helvetica, sans-serif"
        fontSize="0.9rem"
        textAnchor="start"
        fill="#2D3034"
        dy={"0.35em"}
        x={0}
        y={0}
      >
        {" "}
        {label}
      </text>
      <line
        x1={0}
        y1={15}
        x2={textWidth}
        y2={15}
        strokeWidth={4}
        stroke={color}
        strokeLinecap="round"
      />
    </g>
  );
};
