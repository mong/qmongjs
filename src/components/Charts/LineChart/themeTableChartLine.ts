type ThemeTableChartLineTypes = {
  y_axis_tick_number: number;
  y_axis_tick_line_stroke: string;
  x_axis_tick_font_fill: string;
  y_axis_tick_font_fill: string;
  y_axis_label_fill: string;
  y_axis_label_font_family: string;
};

export const themeTableChartLine: ThemeTableChartLineTypes = {
  y_axis_label_fill: "#2D3034",
  y_axis_label_font_family: "arial, Helvetica Neue",
  x_axis_tick_font_fill: "#828586",
  y_axis_tick_font_fill: "#828586",
  y_axis_tick_number: 6,
  y_axis_tick_line_stroke: "#D2D3D4",
};
