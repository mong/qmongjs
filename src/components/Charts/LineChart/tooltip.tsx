import { defaultStyles, Tooltip } from "@visx/tooltip";

import React from "react";
import { DataPoint } from ".";
import { IndicatorData } from "../../IndicatorTable/indicatorvalue";

const tooltipStyles: React.CSSProperties = {
  ...defaultStyles,
  display: "flex",
  flexDirection: "column",
  justifyContent: "space-between",
  alignItems: "center",
  minWidth: 60,
  padding: "10px",
  backgroundColor: "white",
  opacity: 0.8,
  color: "black",
};

type ChartTooltipProps = {
  tooltipOpen: boolean;
  tooltipLeft?: number;
  tooltipTop: number | undefined;
  tooltipData?: DataPoint;
  format?: string;
};

export const LineChartTooltip: React.FC<ChartTooltipProps> = ({
  tooltipData,
  tooltipLeft,
  tooltipOpen,
  tooltipTop,
  format,
}) => {
  return (
    <>
      {tooltipOpen && tooltipData && (
        <Tooltip top={tooltipTop} left={tooltipLeft} style={tooltipStyles}>
          <div
            style={{
              marginBottom: "10px",
              fontSize: "1rem",
              fontWeight: 500,
            }}
          >
            {tooltipData.label}
          </div>
          <div>{tooltipData.year}</div>
          <IndicatorData
            headerStyle={{ fontSize: "1rem" }}
            indicatorData={tooltipData}
            format={format}
          />
        </Tooltip>
      )}
    </>
  );
};
