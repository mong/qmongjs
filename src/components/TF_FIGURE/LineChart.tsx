import React, { useEffect, useRef } from "react";
import { select } from "d3";
import { line_chart } from "../../charts/line_chart";
import { StatisticData } from "../../App";
import { GraphData } from "../main_component";
import { Level } from "./Chart";

export interface Props {
  data: GraphData;
  zoom: boolean;
  showLevel: boolean;
  levels: Level[];
}

export function LineChart(props: Props) {
  const { data, zoom, showLevel, levels } = props;
  const svg_container_ref = useRef<HTMLDivElement>(null);

  useEffect(() => {
    if (!svg_container_ref.current) {
      return;
    }
    let level_visibility = showLevel ? "visible" : "hidden";
    let level = select(svg_container_ref.current);
    level.selectAll("svg .level").style("visibility", level_visibility);
  }, [showLevel]);

  useEffect(() => {
    if (!svg_container_ref.current) return;
    const clientWidth = svg_container_ref.current.clientWidth;
    const svg_props = {
      width: clientWidth,
      height: clientWidth * 0.5,
      margin: { top: 0.05, bottom: 0.1, right: 0.15, left: 0.2 },
      zoom: zoom,
    };

    const nr_svg_children = svg_container_ref.current.childElementCount;
    for (let i = 0; i < nr_svg_children; i++) {
      svg_container_ref.current.removeChild(
        svg_container_ref.current.children[0]
      );
    }
    svg_props.margin = { top: 0.05, bottom: 0.2, right: 0.15, left: 0.05 };
    const filtered_linechart_data: StatisticData[] = data.agg_data.filtered_by_unit.filter(
      (data_by_year: StatisticData) => {
        if (data_by_year.unit_level !== "nation") {
          return !(
            data_by_year.denominator <
              (data.description[0]["min_denominator"] ?? 0) ||
            (data_by_year.dg ?? 0) < 0.6
          ); //|| typeof(data_by_year.dg) === "undefined" )
        } else {
          return true;
        }
      }
    );

    line_chart({
      svg_container: svg_container_ref.current,
      svg_props: svg_props,
      figure_data: filtered_linechart_data,
      levels: levels,
    });
  }, [data, zoom, levels]);

  return <div ref={svg_container_ref} style={{ textAlign: "center" }} />;
}
