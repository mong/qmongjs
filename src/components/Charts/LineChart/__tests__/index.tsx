import { render, screen } from "@testing-library/react";
import { createRef } from "react";
import LineChart, { DataPoint, Props } from "..";
import { buildLevels } from "../../../../test/builders";
import { clockTick } from "../../../../test/clockTick";
import {
  useResizeObserver,
  useLegendItemPosition,
  useTextWidth,
} from "../../../../helpers/hooks";

jest.mock("../../../../helpers/hooks");
jest.mock("../../../../utils/useDelayInitial");

beforeEach(() => {
  const WIDTH = 500;
  (useResizeObserver as jest.Mock).mockReturnValue({
    contentRect: {
      width: WIDTH,
    },
  });
  (useLegendItemPosition as jest.Mock).mockReturnValue({ x: 0, y: 0 });
  (useTextWidth as jest.Mock).mockReturnValue(15);
});

test("shows legend", async () => {
  const d1 = buildDataPoint({ label: "Nasjonalt" });
  const d2 = buildDataPoint({ label: "Ahus" });
  const props = buildProps({ data: [d1, d2] });

  render(<LineChartWithRef {...props} />);

  await clockTick(1500);

  expect(screen.getByText("Nasjonalt")).toBeInTheDocument();
  expect(screen.getByText("Ahus")).toBeInTheDocument();
});

test("shows only one legend item per label", async () => {
  const d1 = buildDataPoint({ label: "Nasjonalt" });
  const d2 = buildDataPoint({ label: "Nasjonalt" });
  const props = buildProps({ data: [d1, d2] });

  render(<LineChartWithRef {...props} />);

  await clockTick(1500);

  expect(screen.getAllByText("Nasjonalt").length).toBe(1);
});

test("Render without levels @250px", async () => {
  const WIDTH = 250;
  (useResizeObserver as jest.Mock).mockReturnValue({
    contentRect: {
      width: WIDTH,
    },
  });
  const { container } = render(
    <LineChartWithRef
      showLevel={false}
      levels={[
        { level: "high", start: 1, end: 0.9 },
        { level: "mid", start: 0.9, end: 0.5 },
        { level: "low", start: 0.5, end: 0 },
      ]}
      tickformat={null}
      data={[
        { label: "a", value: 0.5, year: 2020 },
        { label: "a", value: 0.15, year: 2019 },
        { label: "a", value: 0.3, year: 2018 },
        { label: "a", value: 0.1, year: 2017 },
      ]}
    />
  );

  await clockTick(1500);

  expect(container).toMatchSnapshot();
});

test("Render with levels @500px", async () => {
  const WIDTH = 500;
  (useResizeObserver as jest.Mock).mockReturnValue({
    contentRect: {
      width: WIDTH,
    },
  });
  const { container } = render(
    <LineChartWithRef
      showLevel={true}
      levels={[
        { level: "high", start: 1, end: 0.9 },
        { level: "mid", start: 0.9, end: 0.5 },
        { level: "low", start: 0.5, end: 0 },
      ]}
      tickformat=",.0%"
      data={[
        { label: "test", value: 0.513343, year: 2020 },
        { label: "test", value: 0.15, year: 2019 },
        { label: "test", value: 0.3532, year: 2018 },
        { label: "test", value: 0.124, year: 2017 },
      ]}
    />
  );

  await clockTick(1500);

  expect(container).toMatchSnapshot();
});

test("Render with levels reversed @500px", async () => {
  const WIDTH = 500;
  (useResizeObserver as jest.Mock).mockReturnValue({
    contentRect: {
      width: WIDTH,
    },
  });
  const { container } = render(
    <LineChartWithRef
      showLevel={true}
      levels={[
        { level: "high", start: 0.2, end: 0 },
        { level: "mid", start: 0.4, end: 0.2 },
        { level: "low", start: 1, end: 0.4 },
      ]}
      tickformat=",.3f"
      lastCompleteYear={2019}
      data={[
        { label: "test", value: 0.513343, year: 2020 },
        { label: "test", value: 0.15, year: 2019 },
        { label: "test", value: 0.3532, year: 2018 },
        { label: "test", value: 0.124, year: 2017 },
        { label: "test2", value: 0.13343, year: 2020 },
        { label: "test2", value: 0.5, year: 2019 },
        { label: "test2", value: 0.532, year: 2018 },
        { label: "test2", value: 0.24, year: 2017 },
      ]}
    />
  );

  await clockTick(1500);

  expect(container).toMatchSnapshot();
});

// Helpers
function LineChartWithRef(props: Omit<Props, "svgContainerRef">) {
  const ref = createRef<HTMLDivElement>();
  const WIDTH = 500;
  (useResizeObserver as jest.Mock).mockReturnValue({
    contentRect: {
      width: WIDTH,
    },
  });
  (useLegendItemPosition as jest.Mock).mockReturnValue({ x: 0, y: 0 });

  return <LineChart {...props} svgContainerRef={ref} />;
}

// Builders
function buildDataPoint(overrides: Partial<DataPoint>): DataPoint {
  return {
    label: (Math.random() + 1).toString(36).substring(7),
    year: Math.floor(
      Math.random() * 6 + 2015
    ) /* Random year between 2015 and 2020 */,
    value: Math.random(),
    ...overrides,
  };
}

function buildProps(overrides: Partial<Props>): Omit<Props, "svgContainerRef"> {
  return {
    data: [],
    levels: buildLevels(),
    showLevel: Math.random() < 0.5 /* Random true or false */,
    tickformat: null,
    ...overrides,
  };
}
