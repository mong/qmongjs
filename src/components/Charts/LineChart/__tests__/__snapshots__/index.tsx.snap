// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render with levels @500px 1`] = `
<div>
  <div>
    <div
      style="width: 90%; margin: auto;"
    >
      <svg
        class="lineChart"
        height="322"
        style="background-color: white;"
        width="500"
      >
        <g
          transform="translate(25, 8)"
        >
          <g
            class="legendItem test"
            opacity="1"
            transform="translate(0,0)"
          >
            <text
              dy="0.35em"
              fill="#2D3034"
              font-family="Arial, Helvetica, sans-serif"
              font-size="0.9rem"
              text-anchor="start"
              x="0"
              y="0"
            >
               
              test
            </text>
            <line
              stroke="#4F9A94"
              stroke-linecap="round"
              stroke-width="4"
              x1="0"
              x2="15"
              y1="15"
              y2="15"
            />
          </g>
        </g>
        <g
          transform="translate(25, 38)"
        >
          <g
            class="x-axis"
            fill="none"
            font-family="sans-serif"
            font-size="10"
            text-anchor="middle"
            transform="translate(0, 208)"
          >
            <g
              class="tick"
              opacity="1"
              transform="translate(405.5,0)"
            >
              <text
                dy="0.71em"
                fill="#828586"
                style="font-size: 16.075px;"
                y="21.64"
              >
                2020
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(270.5,0)"
            >
              <text
                dy="0.71em"
                fill="#828586"
                style="font-size: 16.075px;"
                y="21.64"
              >
                2019
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(135.5,0)"
            >
              <text
                dy="0.71em"
                fill="#828586"
                style="font-size: 16.075px;"
                y="21.64"
              >
                2018
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0.5,0)"
            >
              <text
                dy="0.71em"
                fill="#828586"
                style="font-size: 16.075px;"
                y="21.64"
              >
                2017
              </text>
            </g>
          </g>
          <g
            class="y-axis"
            fill="none"
            font-family="sans-serif"
            font-size="10"
            text-anchor="start"
          >
            <text
              fill="#2D3034"
              style="font-size: 15.1px; font-family: arial, Helvetica Neue;"
              transform="rotate(-90)"
              x="-138.66666666666666"
              y="465.74999999999994"
            >
              Andel
            </text>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,208.5)"
            >
              <line
                stroke="#D2D3D4"
                x2="405"
              />
              <text
                dy="0.32em"
                fill="#828586"
                style="font-size: 12.16px;"
                x="408"
              >
                0 %
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,166.9)"
            >
              <line
                stroke="#D2D3D4"
                x2="405"
              />
              <text
                dy="0.32em"
                fill="#828586"
                style="font-size: 12.16px;"
                x="408"
              >
                20 %
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,125.3)"
            >
              <line
                stroke="#D2D3D4"
                x2="405"
              />
              <text
                dy="0.32em"
                fill="#828586"
                style="font-size: 12.16px;"
                x="408"
              >
                40 %
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,83.7)"
            >
              <line
                stroke="#D2D3D4"
                x2="405"
              />
              <text
                dy="0.32em"
                fill="#828586"
                style="font-size: 12.16px;"
                x="408"
              >
                60 %
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,42.099999999999994)"
            >
              <line
                stroke="#D2D3D4"
                x2="405"
              />
              <text
                dy="0.32em"
                fill="#828586"
                style="font-size: 12.16px;"
                x="408"
              >
                80 %
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,0.5)"
            >
              <line
                stroke="#D2D3D4"
                x2="405"
              />
              <text
                dy="0.32em"
                fill="#828586"
                style="font-size: 12.16px;"
                x="408"
              >
                100 %
              </text>
            </g>
          </g>
          <g
            class="levels"
          >
            <rect
              class="level"
              data-testid="level-high"
              fill="#3baa34"
              height="20.80000000000001"
              opacity="0.2"
              width="405"
              x="0"
              y="0"
            />
            <rect
              class="level"
              data-testid="level-mid"
              fill="#fd9c00"
              height="83.19999999999999"
              opacity="0.2"
              width="405"
              x="0"
              y="20.799999999999997"
            />
            <rect
              class="level"
              data-testid="level-low"
              fill="#e30713"
              height="104"
              opacity="0.2"
              width="405"
              x="0"
              y="104"
            />
          </g>
          <g
            class="linesComplete"
          >
            <path
              class="line"
              d="M405,101.224656L270,176.79999999999998L135,134.5344L0,182.208"
              fill="none"
              opacity="1"
              stroke="#4F9A94"
              style="stroke-width: 3; stroke-linejoin: round; stroke-linecap: round; mix-blend-mode: multiply;"
            />
          </g>
          <g
            class="linesIncomplete"
          />
          <g
            class="dots"
          >
            <circle
              class="dot"
              cx="405"
              cy="101.224656"
              fill="#4F9A94"
              opacity="1"
              r="4"
              stroke="#4F9A94"
              style="mix-blend-mode: multiply;"
            />
            <circle
              class="dot"
              cx="270"
              cy="176.79999999999998"
              fill="#4F9A94"
              opacity="1"
              r="4"
              stroke="#4F9A94"
              style="mix-blend-mode: multiply;"
            />
            <circle
              class="dot"
              cx="135"
              cy="134.5344"
              fill="#4F9A94"
              opacity="1"
              r="4"
              stroke="#4F9A94"
              style="mix-blend-mode: multiply;"
            />
            <circle
              class="dot"
              cx="0"
              cy="182.208"
              fill="#4F9A94"
              opacity="1"
              r="4"
              stroke="#4F9A94"
              style="mix-blend-mode: multiply;"
            />
          </g>
        </g>
      </svg>
    </div>
  </div>
</div>
`;

exports[`Render with levels reversed @500px 1`] = `
<div>
  <div>
    <div
      style="width: 90%; margin: auto;"
    >
      <svg
        class="lineChart"
        height="322"
        style="background-color: white;"
        width="500"
      >
        <g
          transform="translate(25, 8)"
        >
          <g
            class="legendItem test"
            opacity="1"
            transform="translate(0,0)"
          >
            <text
              dy="0.35em"
              fill="#2D3034"
              font-family="Arial, Helvetica, sans-serif"
              font-size="0.9rem"
              text-anchor="start"
              x="0"
              y="0"
            >
               
              test
            </text>
            <line
              stroke="#4F9A94"
              stroke-linecap="round"
              stroke-width="4"
              x1="0"
              x2="15"
              y1="15"
              y2="15"
            />
          </g>
          <g
            class="legendItem test2"
            opacity="1"
            transform="translate(0,0)"
          >
            <text
              dy="0.35em"
              fill="#2D3034"
              font-family="Arial, Helvetica, sans-serif"
              font-size="0.9rem"
              text-anchor="start"
              x="0"
              y="0"
            >
               
              test2
            </text>
            <line
              stroke="#90CAF9"
              stroke-linecap="round"
              stroke-width="4"
              x1="0"
              x2="15"
              y1="15"
              y2="15"
            />
          </g>
        </g>
        <g
          transform="translate(25, 38)"
        >
          <g
            class="x-axis"
            fill="none"
            font-family="sans-serif"
            font-size="10"
            text-anchor="middle"
            transform="translate(0, 208)"
          >
            <g
              class="tick"
              opacity="1"
              transform="translate(405.5,0)"
            >
              <text
                dy="0.71em"
                fill="#828586"
                style="font-size: 16.075px;"
                y="21.64"
              >
                2020
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(270.5,0)"
            >
              <text
                dy="0.71em"
                fill="#828586"
                style="font-size: 16.075px;"
                y="21.64"
              >
                2019
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(135.5,0)"
            >
              <text
                dy="0.71em"
                fill="#828586"
                style="font-size: 16.075px;"
                y="21.64"
              >
                2018
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0.5,0)"
            >
              <text
                dy="0.71em"
                fill="#828586"
                style="font-size: 16.075px;"
                y="21.64"
              >
                2017
              </text>
            </g>
          </g>
          <g
            class="y-axis"
            fill="none"
            font-family="sans-serif"
            font-size="10"
            text-anchor="start"
          >
            <text
              fill="#2D3034"
              style="font-size: 15.1px; font-family: arial, Helvetica Neue;"
              transform="rotate(-90)"
              x="-138.66666666666666"
              y="465.74999999999994"
            >
              Antall
            </text>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,208.5)"
            >
              <line
                stroke="#D2D3D4"
                x2="405"
              />
              <text
                dy="0.32em"
                fill="#828586"
                style="font-size: 12.16px;"
                x="408"
              >
                0
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,174.9516129032258)"
            >
              <line
                stroke="#D2D3D4"
                x2="405"
              />
              <text
                dy="0.32em"
                fill="#828586"
                style="font-size: 12.16px;"
                x="408"
              >
                0,1
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,141.40322580645162)"
            >
              <line
                stroke="#D2D3D4"
                x2="405"
              />
              <text
                dy="0.32em"
                fill="#828586"
                style="font-size: 12.16px;"
                x="408"
              >
                0,2
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,107.85483870967741)"
            >
              <line
                stroke="#D2D3D4"
                x2="405"
              />
              <text
                dy="0.32em"
                fill="#828586"
                style="font-size: 12.16px;"
                x="408"
              >
                0,3
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,74.3064516129032)"
            >
              <line
                stroke="#D2D3D4"
                x2="405"
              />
              <text
                dy="0.32em"
                fill="#828586"
                style="font-size: 12.16px;"
                x="408"
              >
                0,4
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,40.75806451612902)"
            >
              <line
                stroke="#D2D3D4"
                x2="405"
              />
              <text
                dy="0.32em"
                fill="#828586"
                style="font-size: 12.16px;"
                x="408"
              >
                0,5
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,7.209677419354856)"
            >
              <line
                stroke="#D2D3D4"
                x2="405"
              />
              <text
                dy="0.32em"
                fill="#828586"
                style="font-size: 12.16px;"
                x="408"
              >
                0,6
              </text>
            </g>
          </g>
          <g
            class="levels"
          >
            <rect
              class="level"
              data-testid="level-high"
              fill="#3baa34"
              height="67.09677419354838"
              opacity="0.2"
              width="405"
              x="0"
              y="140.90322580645162"
            />
            <rect
              class="level"
              data-testid="level-mid"
              fill="#fd9c00"
              height="67.09677419354841"
              opacity="0.2"
              width="405"
              x="0"
              y="73.8064516129032"
            />
            <rect
              class="level"
              data-testid="level-low"
              fill="#e30713"
              height="73.8064516129032"
              opacity="0.2"
              width="405"
              x="0"
              y="0"
            />
          </g>
          <g
            class="linesComplete"
          >
            <path
              class="line"
              d="M270,157.67741935483872L135,89.50709677419354L0,166.4"
              fill="none"
              opacity="1"
              stroke="#4F9A94"
              style="stroke-width: 3; stroke-linejoin: round; stroke-linecap: round; mix-blend-mode: multiply;"
            />
            <path
              class="line"
              d="M270,40.25806451612902L135,29.522580645161273L0,127.48387096774194"
              fill="none"
              opacity="1"
              stroke="#90CAF9"
              style="stroke-width: 3; stroke-linejoin: round; stroke-linecap: round; mix-blend-mode: multiply;"
            />
          </g>
          <g
            class="linesIncomplete"
          >
            <path
              class="line"
              d="M405,35.781703225806446L270,157.67741935483872"
              fill="none"
              opacity="1"
              stroke="#4F9A94"
              style="stroke-width: 3; stroke-dasharray: 10, 10; stroke-linejoin: round; stroke-linecap: round; mix-blend-mode: multiply;"
            />
            <path
              class="line"
              d="M405,163.2363870967742L270,40.25806451612902"
              fill="none"
              opacity="1"
              stroke="#90CAF9"
              style="stroke-width: 3; stroke-dasharray: 10, 10; stroke-linejoin: round; stroke-linecap: round; mix-blend-mode: multiply;"
            />
          </g>
          <g
            class="dots"
          >
            <circle
              class="dot"
              cx="405"
              cy="35.781703225806446"
              fill="#4F9A94"
              opacity="1"
              r="4"
              stroke="#4F9A94"
              style="mix-blend-mode: multiply;"
            />
            <circle
              class="dot"
              cx="270"
              cy="157.67741935483872"
              fill="#4F9A94"
              opacity="1"
              r="4"
              stroke="#4F9A94"
              style="mix-blend-mode: multiply;"
            />
            <circle
              class="dot"
              cx="135"
              cy="89.50709677419354"
              fill="#4F9A94"
              opacity="1"
              r="4"
              stroke="#4F9A94"
              style="mix-blend-mode: multiply;"
            />
            <circle
              class="dot"
              cx="0"
              cy="166.4"
              fill="#4F9A94"
              opacity="1"
              r="4"
              stroke="#4F9A94"
              style="mix-blend-mode: multiply;"
            />
            <circle
              class="dot"
              cx="405"
              cy="163.2363870967742"
              fill="#90CAF9"
              opacity="1"
              r="4"
              stroke="#90CAF9"
              style="mix-blend-mode: multiply;"
            />
            <circle
              class="dot"
              cx="270"
              cy="40.25806451612902"
              fill="#90CAF9"
              opacity="1"
              r="4"
              stroke="#90CAF9"
              style="mix-blend-mode: multiply;"
            />
            <circle
              class="dot"
              cx="135"
              cy="29.522580645161273"
              fill="#90CAF9"
              opacity="1"
              r="4"
              stroke="#90CAF9"
              style="mix-blend-mode: multiply;"
            />
            <circle
              class="dot"
              cx="0"
              cy="127.48387096774194"
              fill="#90CAF9"
              opacity="1"
              r="4"
              stroke="#90CAF9"
              style="mix-blend-mode: multiply;"
            />
          </g>
        </g>
      </svg>
    </div>
  </div>
</div>
`;

exports[`Render without levels @250px 1`] = `
<div>
  <div>
    <div
      style="width: 90%; margin: auto;"
    >
      <svg
        class="lineChart"
        height="322"
        style="background-color: white;"
        width="500"
      >
        <g
          transform="translate(25, 8)"
        >
          <g
            class="legendItem a"
            opacity="1"
            transform="translate(0,0)"
          >
            <text
              dy="0.35em"
              fill="#2D3034"
              font-family="Arial, Helvetica, sans-serif"
              font-size="0.9rem"
              text-anchor="start"
              x="0"
              y="0"
            >
               
              a
            </text>
            <line
              stroke="#4F9A94"
              stroke-linecap="round"
              stroke-width="4"
              x1="0"
              x2="15"
              y1="15"
              y2="15"
            />
          </g>
        </g>
        <g
          transform="translate(25, 38)"
        >
          <g
            class="x-axis"
            fill="none"
            font-family="sans-serif"
            font-size="10"
            text-anchor="middle"
            transform="translate(0, 208)"
          >
            <g
              class="tick"
              opacity="1"
              transform="translate(405.5,0)"
            >
              <text
                dy="0.71em"
                fill="#828586"
                style="font-size: 16.075px;"
                y="21.64"
              >
                2020
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(270.5,0)"
            >
              <text
                dy="0.71em"
                fill="#828586"
                style="font-size: 16.075px;"
                y="21.64"
              >
                2019
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(135.5,0)"
            >
              <text
                dy="0.71em"
                fill="#828586"
                style="font-size: 16.075px;"
                y="21.64"
              >
                2018
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0.5,0)"
            >
              <text
                dy="0.71em"
                fill="#828586"
                style="font-size: 16.075px;"
                y="21.64"
              >
                2017
              </text>
            </g>
          </g>
          <g
            class="y-axis"
            fill="none"
            font-family="sans-serif"
            font-size="10"
            text-anchor="start"
          >
            <text
              fill="#2D3034"
              style="font-size: 15.1px; font-family: arial, Helvetica Neue;"
              transform="rotate(-90)"
              x="-138.66666666666666"
              y="465.74999999999994"
            >
              Antall
            </text>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,208.5)"
            >
              <line
                stroke="#D2D3D4"
                x2="405"
              />
              <text
                dy="0.32em"
                fill="#828586"
                style="font-size: 12.16px;"
                x="408"
              >
                0
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,173.24576271186442)"
            >
              <line
                stroke="#D2D3D4"
                x2="405"
              />
              <text
                dy="0.32em"
                fill="#828586"
                style="font-size: 12.16px;"
                x="408"
              >
                0,1
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,137.99152542372883)"
            >
              <line
                stroke="#D2D3D4"
                x2="405"
              />
              <text
                dy="0.32em"
                fill="#828586"
                style="font-size: 12.16px;"
                x="408"
              >
                0,2
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,102.73728813559322)"
            >
              <line
                stroke="#D2D3D4"
                x2="405"
              />
              <text
                dy="0.32em"
                fill="#828586"
                style="font-size: 12.16px;"
                x="408"
              >
                0,3
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,67.4830508474576)"
            >
              <line
                stroke="#D2D3D4"
                x2="405"
              />
              <text
                dy="0.32em"
                fill="#828586"
                style="font-size: 12.16px;"
                x="408"
              >
                0,4
              </text>
            </g>
            <g
              class="tick"
              opacity="1"
              transform="translate(0,32.22881355932202)"
            >
              <line
                stroke="#D2D3D4"
                x2="405"
              />
              <text
                dy="0.32em"
                fill="#828586"
                style="font-size: 12.16px;"
                x="408"
              >
                0,5
              </text>
            </g>
          </g>
          <g
            class="levels"
          />
          <g
            class="linesComplete"
          >
            <path
              class="line"
              d="M405,31.72881355932202L270,155.11864406779662L135,102.23728813559322L0,172.74576271186442"
              fill="none"
              opacity="1"
              stroke="#4F9A94"
              style="stroke-width: 3; stroke-linejoin: round; stroke-linecap: round; mix-blend-mode: multiply;"
            />
          </g>
          <g
            class="linesIncomplete"
          />
          <g
            class="dots"
          >
            <circle
              class="dot"
              cx="405"
              cy="31.72881355932202"
              fill="#4F9A94"
              opacity="1"
              r="4"
              stroke="#4F9A94"
              style="mix-blend-mode: multiply;"
            />
            <circle
              class="dot"
              cx="270"
              cy="155.11864406779662"
              fill="#4F9A94"
              opacity="1"
              r="4"
              stroke="#4F9A94"
              style="mix-blend-mode: multiply;"
            />
            <circle
              class="dot"
              cx="135"
              cy="102.23728813559322"
              fill="#4F9A94"
              opacity="1"
              r="4"
              stroke="#4F9A94"
              style="mix-blend-mode: multiply;"
            />
            <circle
              class="dot"
              cx="0"
              cy="172.74576271186442"
              fill="#4F9A94"
              opacity="1"
              r="4"
              stroke="#4F9A94"
              style="mix-blend-mode: multiply;"
            />
          </g>
        </g>
      </svg>
    </div>
  </div>
</div>
`;
